Contributing guidelines update (#169)
merely-useful/merely-useful.github.io: .github/CONTRIBUTING.md

    "Should we implement this decision over the stated objections?"
    If a majority votes "yes" the proposal is implemented.
    Otherwise, the proposal is returned to the sponsor for further work.

## Building and Previewing

**Note:** we are using `book` as the master (published) branch because of the
way GitHub handles organization repositories, domain names, and the generation
of the website.

**Setup to locally preview the book**:

1.  Install R (we recommend that you also install and use the RStudio IDE).

1.  Open RStudio by clicking on the `merely-useful.github.io.Rproj` file (if not
through RStudio, then open an R console in the location of the Merely Useful
repo) and install the dependencies by typing in the console:

    ```r
    # Make sure you have remotes installed
    # install.packages("remotes")
    remotes::install_deps()
    ```

**Build and preview from the command line**:

1.  `make html` and then open `_book/index.html`.
1.  `make pdf` or `make epub` to build PDF and EPUB versions (also in the
`_book` folder).

Or via RStudio:

1. When in the R Project (opened via the `.Rproj` file), use the key bindings `Ctrl-Shift-B` to build the `html` output.

> ### Continuous Building using Travis CI

Reviewer: Maybe move this section last since it is more of an FYI than something contributors will need to worry about? It could be its on CI section instead of under building.

Final building of the website is done via Travis CI. In order to get Travis set
up to push the final generated book to the `master` branch, a GitHub Personal
Access Token (PAT) must be added. This PAT can be generated and assigned to
Travis with the following steps:

1. Create a [PAT](https://help.github.com/en/articles/creating-a-personal-access-token-for-the-command-line)
for your account on GitHub (make sure to enable the "repo" scope so that using
this token will enable writing to your GitHub repos) and copy the token to your
clipboard.
1. Go to https://travis-ci.org/USER/REPO/settings replacing `USER` with your
GitHub ID and `REPO` with the name of the forked repository.
1. Under the section "Environment Variables", type `GITHUB_TOKEN` in the "Name"
text box and paste your personal access token into the "Value" text box.
1. The `deploy` Travis commands found in [`.travis.yml`](.travis.yml) are then
used to access this GitHub PAT.

## Workflow for Adding Content

- Use the `book` branch as the main "development" branch (**not `master`**).
- Since `book` is "protected", to add to the repo you must:

>     1. Submit a PR of a branch from their fork of the repo

Reviewer: ```suggestion
    1. Submit a PR of a branch from your fork of the repo
```

- If making a suggested change to the actual text, please use the
["Insert a suggestion"](https://help.github.com/en/articles/commenting-on-a-pull-request#adding-line-comments-to-a-pull-request)
    - For the **"reviewee"**, please accept those changes. GitHub will insert the
    suggestion directly as a commit, thus giving attribution to the reviewer. Note:
    This does not work for multi-line suggestions edits (though you can add more
    lines in the suggestion if desired).
- For individual chapter reviews, look for consistency, content, clarity of
material with exercises. Don't worry too much about grammar appearance, structure,
etc. This will be done at the end stage.

## Specific Content Guidelines

1.  bookdown's behavior is controlled by the metadata in `_bookdown.yml` and in the YAML header of `index.Rmd`.
    Please do not add YAML headers to other `.Rmd` files, since it is ignored when building the entire book.

1.  `_bookdown.yml` contains an ordered list of R Markdown chapters and appendices.
    This list is repeated in `Makefile`.

1.  The first element of each chapter must be a level-1 heading.
    Anything before this heading is silently ignored.

> 1.  The ID for each chapter and section heading is set by putting `{#some-label}` on the heading line.

Reviewer: Is bookdown not using [Pandoc's default with auto_identifiers](https://pandoc.org/MANUAL.html#extension-auto_identifiers) to generate these automatically?

Reviewer: Or do we do this to ensure no duplicates? If so, should we preface with the course (py/r/rse)?

Author: This text was there from @gvwilson. bookdown does use pandoc auto identifying, but I think in this case it's to help with duplicates and to distinguish course sections, as you say. I'll add the "py", "r", etc comment.

    Please use the stem of the file's name in labels, i.e., start all labels for chapter `stuff.Rmd` with the word `stuff`.

1.  Each chapter starts with a list of questions and a list of learning objectives and ends with a list of key points.

>     These lists are kept in Markdown files in the `questions`, `objectives`, and `keypoints` folders respectively

Reviewer: ```suggestion
    These lists are kept in plain Markdown files in the `questions`, `objectives`, and `keypoints` folders respectively
```
Just for clarity, i got confused for a second

1.  The last line in each chapter includes the file `links.md`, which gives symbolic names to all the external links used in the book.
    Please always use `[text][label]` to refer to links and put the URL itself in this file to ensure consistency between chapters.

1.  Please use the free desktop version of [draw.io](https://www.draw.io/) to draw diagrams.
    Save the source as `figures/stem/stem.drawio` and export drawings as all of PDF, PNG, and SVG.

1.  Since LaTeX doesn't understand SVG images, hand-drawn diagrams are included as follows:
    -   Create an R code block with the header `{r stem-label, echo=FALSE, fig.cap="Some Caption"}`, which means:
        -   Give this figure the ID `fig:stem-label`.
            (Bookdown automatically puts `fig:` in front of figure labels,
            though it doesn't prefix section labels with anything.)
        -   Don't show the R code used to load the image, just its output (i.e., the image).
        -   Give the figure the specified caption.
    -   Use `insert_graphic("figures/stem/filename.ext")` to include your image. If you are using a `.pdf` image, make sure that you have the same file but as a `.svg` as well. 

> 1.  Use `@Name1234` to refer to bibliography entries.

Reviewer: Don't we want `[@Name1234]` so that the citations are surrounded by parentheses? And leave out `[]` only when the citation is truly part of the sentence (the "in-text" [example here](https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)). Mention `;` for multiple references or just include the link I just posted.

Author: I'm not always sure how the [ works, since it depends on the style. I included mention of the [, but don't expand on it.

1.  Use `\@ref(label)` to refer to labels for sections and figures.
    Note that:
    -   This only inserts the section number, not the words `Chapter`, `Appendix`, `Section`, or `Figure`.
    -   Those are round parentheses, not curly braces.
    -   If the figure's chunk ID is `stem-label`, use `fig:stem-label` to refer to it.

1.  Glossary entries are in `gloss.md`, which is plain Markdown rather than R Markdown.
    -   To refer to a glossary entry, use a direct link of the form `[text]``(glossary.html#term-label)`.
    -   Glossary definitions are set in bold and use an HTML anchor tag to provide the ID.
        We should find a more elegant way to do this.

1.  Many code fragments are accompanied by an HTML comment that contains `src="stem/filename.ext"`.
    This identifies the path to that file within the `./src` folder.

>     We should find a more elegant way to handle these references.

Reviewer: For this and the bullet above? Are the current ways robust? If so, do they need to be more elegant?

Author: I don't actually know what this part describes, as this was for guidelines before we decided on bookdown. @gvwilson, any comment? Do we need this section?

Reviewer 2: This point can be removed now.

1.  There are also some HTML comments containing the word `noindent` left over from formatting with an earlier template.
    These were used to prevent indentation of the first line of continuation paragraphs.
    We will find a more elegant way to handle this as we get closer to production.
    

>     -   Do *not* use `View` in your R snippets,

Reviewer: Does this belong to the bullet above, or should it be its own point?

        as Travis will fail when it tries to launch a viewer from the command line.
    

## Using Python Code Chunks

> R users should be already familiar with using R Markdown. The process for using

Reviewer: ```suggestion
The process for inserting Python code chunks in R Markdown is the same as for inserting R code chunks. Simply replace `{r}` with `{python}` on the first line of the fenced code chunk: 
```

> Python in R Markdown is familiar straight forward. When you want to include some

Reviewer: ```suggestion

```

> code and its output, use code chunks like:

Reviewer: ```suggestion

```

    
    ```{python}

>     python code ...

Reviewer: ```suggestion
    # Python chunks will also be syntax highlighted
    for num in range(5):
        print(num)
```

Reviewer: I guess this maybe is unnecessary... i was thinking that this section was in the book so it would be nice to show higlighting, but it is just in the contributing which won't recognize `{lang}` anyways. Oh well

    ```

Code chunks have
[options](https://www.rstudio.com/wp-content/uploads/2015/03/rmarkdown-reference.pdf)
that apply to both R and Python.
    
## When Using Other Software Packages

Travis needs to know what packages to install to build the website:

- For R packages, it is very easy. When in the Merely Useful R Project (by
opening the `merely-useful.github.io.Rproj` file), type
`usethis::use_package("packagename")` in the R Console. This will add the
package dependency to the `DESCRIPTION` file under the imports section.

- For Python packages you'll need to add the package to the `.travis.yml` under 
the `addons` section:

```yaml
addons:
  apt:
    packages:
    - python3
    # Add more packages below here
    # and looks like "python3-packagename"
    # e.g.:
    - python3-numpy
```

> ## Guide for code and text styling

Reviewer: Maybe move this above the previous section and then move down the Travis section I commented on above to the bottom. Then both the travis sections are at the end.
